File: bench/src/ts/fnptr.ts
type Op = (x: number) => number;
function add1(x: number): number { return x+1; }
function mul3(x: number): number { return x*3; }
function apply(op: Op, x: number): number { return op(x); }
const a: Op = add1, b: Op = mul3;
let sum=0;
for(let i=0;i<8000000;i++){ if(i%2===0) sum+=apply(a,i); else sum+=apply(b,i); }
process.exit(sum % 256);

export {};

File: bench/src/rust/fnptr.rs
type Op = fn(i64)->i64;
fn add1(x:i64)->i64{ x+1 }
fn mul3(x:i64)->i64{ x*3 }
fn apply(op:Op, x:i64)->i64{ op(x) }
fn main(){
  let a:Op=add1; let b:Op=mul3; let mut sum=0u64;
  for i in 0u64..8000000 { if i%2==0 { sum+=apply(a,i as i64) as u64 } else { sum+=apply(b,i as i64) as u64 } }
  std::process::exit((sum%256) as i32);
}
